### PR TITLE
Add some check and retry for diagnosing flaky tests

### PR DESCRIPTION
**Pull request by GangWang01**
### Problem
#4618 - some flaky tests in CI

### Solution
Add some check and log to help on diagnostic.
For network related issue, make it retry some times.

### Checks:
- [ ] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: test/Microsoft.TemplateSearch.Common.UnitTests/NuGetMetadataSearchProviderTests.cs [hunk 1]
```diff
@@ -334,6 +334,31 @@ private string GenerateLocalCache()
             return targetPath;
         }
 
+        private async Task<IReadOnlyList<SearchResult>> AttemptSearchAsync(int count, int intervalInSeconds, Func<TemplatePackageSearchData, IReadOnlyList<ITemplateInfo>> filter, TemplateSearchCoordinator searchCoordinator)
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateSearch.Common.UnitTests/NuGetMetadataSearchProviderTests.cs:337]: `TimeSpan` might be better type for a duration argument

File: test/Microsoft.TemplateSearch.Common.UnitTests/TemplateSearchCacheReaderTests.cs [hunk 2]
```diff
@@ -171,5 +171,30 @@ public void CanReadSearchMetadata_V2_E2E()
             Assert.Equal(2, ((ITemplateInfo)templateToTest).PostActions.Count);
             Assert.Equal(new[] { postAction1, postAction2 }, ((ITemplateInfo)templateToTest).PostActions);
         }
+
+        private async Task<string> AttemptGetSearchFileAsync(int count, int intervalInSeconds, NuGetMetadataSearchProvider searchProvider)
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateSearch.Common.UnitTests/TemplateSearchCacheReaderTests.cs:175]: `AttemptSearchAsync` and `AttemptGetSearchFileAsync` can be unified into single helper function taking a functor and nonretry exception type

File: test/dotnet-new3.UnitTests/DotnetNewInstall.cs [hunk 3]
```diff
@@ -615,7 +615,10 @@ public void CannotInstallTemplateWithoutMandatoryConfig()
         {
             var home = TestUtils.CreateTemporaryFolder("Home");
             string invalidTemplatePath = TestUtils.GetTestTemplateLocation("Invalid/MissingMandatoryConfig");
+            var assets = new string[] { Path.Combine(invalidTemplatePath, "template.json") };
+            TestUtils.EnsureTestAssetsAvailable(assets, _log);
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/DotnetNewInstall.cs:619]: as discussed, this fix and below will probably not help. 
Let's enable debug and add more logging to try to get more info.

Comment (GangWang01) [reply test/dotnet-new3.UnitTests/DotnetNewInstall.cs:619]: ah, my fault. I will add more debug info instead.

**Review by vlada-shubina — APPROVED** [review]
Great work, thank you!

**Pull request merged**